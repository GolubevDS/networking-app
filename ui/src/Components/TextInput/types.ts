import type { FC, SVGProps } from 'react';

/**
 * Props for the TextInput component.
 *
 * @typedef {object} TextInputProps
 * @property {boolean} [clearable=false] - Whether the input should have a clear button.
 * @property {React.FC<React.SVGProps<SVGElement>>} [Icon] - An optional icon to display in the input.
 * @property {string} [placeholder] - Placeholder text to display in the input.
 * @property {string} [value] - Initial value for the input
 */
export interface TextInputProps {
	clearable?: boolean;
	Icon?: FC<SVGProps<SVGElement>>;
	placeholder?: string;
	value?: string;
}
