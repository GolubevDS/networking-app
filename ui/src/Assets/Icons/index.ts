export { default as Calendar } from './Calendar.svg';
export { default as Camera }   from './Camera.svg';
export { default as Contacts } from './Contacts.svg';
export { default as Close }    from './Close.svg';
export { default as Logo }     from './Logo.svg';
export { default as Search }   from './Search.svg';
export { default as Settings } from './Settings.svg';

