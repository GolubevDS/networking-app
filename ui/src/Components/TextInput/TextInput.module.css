.root {
	--font-size:     var(--font-size-body);
	--line-height:   var(--font-line-height-body);
	--icon-size:     var(--line-height);
	border:          var(--border-width) solid var(--color-border);
	transition:      var(--transition);
	border-radius:   6px;
	display:         flex;
	justify-content: flex-start;
	align-items:     center;
	width:           50rem;
}

.root:focus-within, .root:has(.textField:not(:placeholder-shown)) {
	border-color: var(--color-border-focus);
}

svg {
	height: var(--icon-size);
	width:  var(--icon-size);
}

.label {
	padding:         .75em 0 .75em 1em;
	cursor:          text;
	flex-grow:       1;
	display:         flex;
	justify-content: flex-start;
	align-items:     center;
}

.label > svg {
	fill:         var(--color-border);
	margin-right: .5em;
	transition:   var(--transition);
}

.label:focus-within > svg, .label:has(.textField:not(:placeholder-shown)) > svg {
	fill: var(--color-border-focus);
}

.textField {
	font-size:   var(--font-size);
	line-height: var(--line-height);
	width:       100%;
}

::placeholder {
	color: var(--color-text-placeholder);
}

.clearInput {
	fill:         var(--color-text);
	transition:   var(--transition);
	cursor:       pointer;
	margin-left:  .5em;
	margin-right: 1em;
}

.clearInput:hover {
	fill: var(--primary-hover);
}
