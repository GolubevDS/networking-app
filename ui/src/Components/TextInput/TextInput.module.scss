.root {
	--icon-size:     var(--font-line-m);
	border:          var(--border-width) solid var(--border-color);
	transition:      var(--transition);
	border-radius:   6px;
	display:         flex;
	justify-content: flex-start;
	align-items:     center;
	width:           50rem;
}

.root:focus-within, .root:has(.textField:not(:placeholder-shown)) {
	border-color: var(--border-focus-color);
}

svg {
	height: var(--icon-size);
	width:  var(--icon-size);
}

.label {
	padding:         .75em 0 .75em 1em;
	cursor:          text;
	flex-grow:       1;
	display:         flex;
	justify-content: flex-start;
	align-items:     center;

	& > svg {
		fill:         var(--border-color);
		margin-right: .5em;
		transition:   var(--transition);
	}
}

.label:focus-within > svg, .label:has(.textField:not(:placeholder-shown)) > svg {
	fill: var(--border-focus-color);
}

.textField {
	font:  var(--font-body);
	color: var(--text-color);
	width: 100%;
}

::placeholder {
	color: var(--placeholder-color);
}

.clearInput {
	fill:         var(--text-color);
	transition:   var(--transition);
	cursor:       pointer;
	margin-left:  .5em;
	margin-right: 1em;

	&:hover {
		fill: var(--primary-hover-color);
	}
}
