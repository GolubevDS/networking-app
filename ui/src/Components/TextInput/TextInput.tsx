import React, { useState }  from 'react';
import type { ChangeEvent } from 'react';

import { Close } from '~Assets/Icons';

import styles                  from './TextInput.module.css';
import type { TextInputProps } from './types';

/** TextInput allows users to enter text into an input field. */
export const TextInput = ({
	clearable,
	Icon,
	placeholder = '',
	value: defaultValue = '',
}: TextInputProps): JSX.Element => {
	const [value, setValue] = useState(defaultValue);

	const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
		setValue(event.currentTarget.value);
	};

	const handleClearInput = () => {
		setValue('');
	};

	return (
		<div className={styles.root}>
			<label className={styles.label}>
				{Icon ? <Icon data-testid="icon" /> : null}
				<input
					className={styles.textField}
					placeholder={placeholder}
					type="text"
					value={value}
					onChange={handleChange}
				/>
			</label>
			{clearable && value ? (
				<Close
					className={styles.clearInput}
					data-testid="clear-button"
					onClick={handleClearInput}
				/>
			) : null}
		</div>
	);
};
