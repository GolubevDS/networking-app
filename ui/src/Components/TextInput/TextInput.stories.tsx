import type { ComponentMeta } from '@storybook/react';
import React                  from 'react';

import * as icons from '~Assets/Icons';

import { TextInput }           from './TextInput';
import type { TextInputProps } from './types';

export default {
	title:     'Text input',
	component: TextInput,
	argTypes:  {
		Icon: {
			options: Object.keys(icons),
			mapping: icons,
			control: {
				defaultValue: null,
				description:  'Icon',
				type:         'select',
				labels:       Object.keys(icons),
			},
		},
	},
} as ComponentMeta<typeof TextInput>;

export const Default = (args: TextInputProps) => <TextInput {...args} />;
