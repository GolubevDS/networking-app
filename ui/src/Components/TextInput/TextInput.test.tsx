import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import React                 from 'react';

import { Search } from '~Assets/Icons';

import { TextInput } from './TextInput';

describe('TextInput', () => {
	test('renders correctly', () => {
		const { container } = render(
			<TextInput
				clearable
				Icon={Search}
				placeholder="Search here..."
				value="Hello, world!"
			/>,
		);
		expect(container).toMatchSnapshot();
	});

	test('allows users to enter text', () => {
		const { getByPlaceholderText } = render(<TextInput placeholder="Search here..." />);
		const input = getByPlaceholderText('Search here...') as HTMLInputElement;

		fireEvent.change(input, { target: { value: 'Hello, world!' } });
		expect(input.value).toBe('Hello, world!');
	});

	test('displays an icon if provided', () => {
		const { getByTestId } = render(<TextInput Icon={Search} />);
		const icon = getByTestId('icon');

		expect(icon).toBeInTheDocument();
	});

	test('does not display an icon if not provided', () => {
		const { queryByTestId } = render(<TextInput />);
		const icon = queryByTestId('icon');

		expect(icon).not.toBeInTheDocument();
	});

	test('clears the input when the clear button is clicked', () => {
		const { getByPlaceholderText, getByTestId } = render(<TextInput clearable placeholder="Search here..." />);
		const input = getByPlaceholderText('Search here...') as HTMLInputElement;

		fireEvent.change(input, { target: { value: 'Hello, world!' } });
		expect(input.value).toBe('Hello, world!');

		const clearButton = getByTestId('clear-button');

		fireEvent.click(clearButton);
		expect(input.value).toBe('');
	});
});
