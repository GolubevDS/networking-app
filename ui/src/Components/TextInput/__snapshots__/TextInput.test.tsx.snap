// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders correctly 1`] = `
<div>
  <div
    class="root"
  >
    <label
      class="label"
    >
      <svg
        data-testid="icon"
      />
      <input
        class="textField"
        placeholder="Search here..."
        type="text"
        value="Hello, world!"
      />
    </label>
    <svg
      class="clearInput"
      data-testid="clear-button"
    />
  </div>
</div>
`;
